src/main/java/com/fasterxml/jackson/databind/deser/std/CollectionDeserializer.java

```diff
@@ -353,12 +353,16 @@ protected Collection<Object> _deserializeFromArray(JsonParser p, Deserialization
                     if (_skipNullValues) {
                         continue;
                     }
-                    value = _nullProvider.getNullValue(ctxt);
-                } else if (_valueTypeDeserializer == null) {
-                    value = _valueDeserializer.deserialize(p, ctxt);
+                    value = null;
                 } else {
-                    value = _valueDeserializer.deserializeWithType(p, ctxt, _valueTypeDeserializer);
+                    value = _deserializeRawContentValue(p, ctxt);
+                }
+
+                if (value == null) {
+                    value = _nullProvider.getNullValue(ctxt);
                 }
+
+                // _skipNullValues is checked by _tryToAddNull.
                 if (value == null) {
```

Comment:
This part can be moved within previous block, after `value = _nullProvider.getNullValue(ctxt);`

Reply:
Fixed
aa4dbab